gax-grpc/src/main/java/com/google/api/gax/grpc/ChannelPool.java

```diff
@@ -207,26 +256,214 @@ public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedE
     return isTerminated();
   }
 
+  /** Scheduling loop. */
+  private ScheduledFuture<?> scheduleNextRefresh() {
+    long delayPeriod = REFRESH_PERIOD.toMillis();
+    long jitter = (long) ((Math.random() - 0.5) * JITTER_PERCENTAGE * delayPeriod);
+    long delay = jitter + delayPeriod;
+    return channelRefreshExecutorService.schedule(
+        () -> {
+          scheduleNextRefresh();
+          refresh();
+        },
+        delay,
+        TimeUnit.MILLISECONDS);
+  }
+
   /**
-   * Performs a simple round-robin on the list of {@link ManagedChannel}s in the {@code channels}
-   * list.
+   * Replace all of the channels in the channel pool with fresh ones. This is meant to mitigate the
+   * hourly GFE disconnects by giving clients the ability to prime the channel on reconnect.
    *
-   * @return A {@link ManagedChannel} that can be used for a single RPC call.
+   * <p>This is done on a best effort basis. If the replacement channel fails to construct, the old
+   * channel will continue to be used.
    */
-  private ManagedChannel getNextChannel() {
-    return getChannel(indexTicker.getAndIncrement());
+  @InternalApi("Visible for testing")
+  void refresh() {
+    // Note: synchronization is necessary in case refresh is called concurrently:
+    // - thread1 fails to replace a single entry
+    // - thread2 succeeds replacing an entry
+    // - thread1 loses the race to replace the list
+    // - then thread2 will shut down channel that thread1 will put back into circulation (after it
+    //   replaces the list)
+    synchronized (entryWriteLock) {
+      ArrayList<Entry> newEntries = new ArrayList<>(entries.get());
+
+      for (int i = 0; i < newEntries.size(); i++) {
+        try {
+          newEntries.set(i, new Entry(channelFactory.createSingleChannel()));
+        } catch (IOException e) {
+          LOG.log(Level.WARNING, "Failed to refresh channel, leaving old channel", e);
+        }
+      }
+
+      ImmutableList<Entry> replacedEntries = entries.getAndSet(ImmutableList.copyOf(newEntries));
+
+      // Shutdown the channels that were cycled out. This will either be the channels we just
+      // refreshed or in case of a race, the channels that the other thread set.
```

Comment:
Since there's no longer a race after synchronization,
```suggestion
      // Shutdown the channels that were cycled out.
```